src/hotspot/share/opto/divnode.cpp

```diff
@@ -1257,6 +1386,152 @@ const Type* ModLNode::Value(PhaseGVN* phase) const {
+  if(in(0) && in(0)->is_top()) {
+    return nullptr;
+  }
+
+  // Get the modulus
+  const Type* t = phase->type(in(2));
+  if(t == Type::TOP) {
+    return nullptr;
+  }
+  const TypeInt* ti = t->is_int();
+
+  // Check for useless control input
+  // Check for excluding mod-zero case
+  if (in(0) && (ti->_hi < 0 || ti->_lo > 0)) {
+    set_req(0, nullptr);        // Yank control input
+    return this;
+  }
+
+  if(!ti->is_con()) {
+    return nullptr;
+  }
+  juint con = ti->get_con();
+  const Type* u = phase->type(in(1));
+  if (!u->isa_int() || u->is_int()->is_con()) {
+    return nullptr;
+  }
+
+  // See if we are MOD'ing by 2^k
+  if (is_power_of_2(con)) {
+    return new AndINode(in(1), phase->intcon(con - 1));
+  }
+  // TODO: This can be calculated directly, see https://arxiv.org/abs/1902.01961
+  Node* q = transform_int_udivide(phase, in(1), con);
+  if (q == nullptr) {
+    return nullptr;
+  }
+  q = phase->transform(q);
+  Node* mul = phase->transform(new MulINode(q, phase->intcon(con)));
+  return new SubINode(in(1), mul);
+}
+
+//------------------------------Value------------------------------------------
+const Type* UModINode::Value(PhaseGVN* phase) const {
+  // Either input is TOP ==> the result is TOP
+  const Type* t1 = phase->type( in(1) );
+  const Type* t2 = phase->type( in(2) );
+  if(t1 == Type::TOP || t2 == Type::TOP) {
+    return Type::TOP;
+  }
+
+  // x % x = 0, x % 1 = 0
+  if (in(1) == in(2) || t2 == TypeInt::ONE) {
+    return TypeInt::ZERO;
+  }
+
+  const TypeInt* i1 = t1->is_int();
+  const TypeInt* i2 = t2->is_int();
+
+  // constant fold
+  if (i1->is_con() && i2->is_con()) {
+    return TypeInt::make(juint(i1->get_con()) % juint(i2->get_con()));
+  }
+
+  return TypeInt::INT;
+}
+
+//=============================================================================
+//------------------------------Idealize---------------------------------------
+Node* UModLNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  // Check for dead control input
+  if(in(0) && remove_dead_region(phase, can_reshape)) {
+    return this;
+  }
+  // Don't bother trying to transform a dead node
+  if(in(0) && in(0)->is_top()) {
+    return nullptr;
+  }
+
+  // Get the modulus
+  const Type* t = phase->type(in(2));
+  if(t == Type::TOP) {
+    return nullptr;
+  }
+  const TypeLong* ti = t->is_long();
+
+  // Check for useless control input
+  // Check for excluding mod-zero case
+  if (in(0) && (ti->_hi < 0 || ti->_lo > 0)) {
+    set_req(0, nullptr);        // Yank control input
+    return this;
+  }
+
+  if(!ti->is_con()) {
+    return nullptr;
+  }
+  julong con = ti->get_con();
+  const Type* u = phase->type(in(1));
+  if (!u->isa_long() || u->is_long()->is_con()) {
+    return nullptr;
+  }
+
+  // See if we are MOD'ing by 2^k
+  if (is_power_of_2(con)) {
+    return new AndLNode(in(1), phase->longcon(con - 1));
+  }
+  Node* q = transform_long_udivide(phase, in(1), con);
+  if (q == nullptr) {
```

Comment:
Maybe assert `!Matcher::match_rule_supported(Op_UMulHiL)`, if that is the only case it should happen?